opm/simulators/wells/WellBhpThpCalculator.cpp

```diff
@@ -839,6 +866,114 @@ bruteForceBracket(const std::function<double(const double)>& eq,
+        const auto ipr = getFloIPR(well_state, well, summaryState);
+        return bhp.dflo + 1/ipr.second >= 0; 
+    }                  
+}
+
+std::optional<double> WellBhpThpCalculator::
+estimateStableBhp(const WellState& well_state,
+                  const Well& well,
+                  const std::vector<double>& rates,
+                  const double rho,
+                  const SummaryState& summaryState,
+                  DeferredLogger& deferred_logger) const
+{   
+    // Given a *converged* well_state with ipr, estimate bhp of the stable solution 
+    const auto& controls = well.productionControls(summaryState);
+    const double thp = well_.getTHPConstraint(summaryState);
+    const auto& table = well_.vfpProperties()->getProd()->getTable(controls.vfp_table_number);
+
+    const double aqua = rates[BlackoilPhases::Aqua];
+    const double liquid = rates[BlackoilPhases::Liquid];
+    const double vapour = rates[BlackoilPhases::Vapour];
+    double flo = detail::getFlo(table, aqua, liquid, vapour);
+    double wfr, gfr;
+    if (well_.useVfpExplicit() || -flo < table.getFloAxis().front()) {
+        wfr =  well_.vfpProperties()->getExplicitWFR(controls.vfp_table_number, well_.indexOfWell());
+        gfr = well_.vfpProperties()->getExplicitGFR(controls.vfp_table_number, well_.indexOfWell());
+    } else {
+        wfr = detail::getWFR(table, aqua, liquid, vapour);  
+        gfr = detail::getGFR(table, aqua, liquid, vapour);   
+    }
+    if (wfr <= 0.0 && gfr <= 0.0) {
+        // warning message
+    }
+    auto ipr = getFloIPR(well_state, well, summaryState);
+    
+    const double vfp_ref_depth = well_.vfpProperties()->getProd()->getTable(controls.vfp_table_number).getDatumDepth();
+
+    const double dp_hydro = wellhelpers::computeHydrostaticCorrection(well_.refDepth(), vfp_ref_depth, 
+                                                                      rho, well_.gravity());
+    if (ipr.first <= 0.0) {
+        // error message
+    }
+    auto bhp_adjusted = [this, &thp, &dp_hydro](const double bhp) {
+           return bhp - dp_hydro + getVfpBhpAdjustment(bhp, thp);
+       };
+    const auto retval = detail::intersectWithIPR(table, thp, wfr, gfr, well_.getALQ(well_state), ipr.first, ipr.second, bhp_adjusted);
+    if (retval.has_value()) {
+        // returned pair is (flo, bhp)
+        return retval.value().second;
+    } else {
+        return std::nullopt;
+    }
+}    
+
+std::pair<double, double> WellBhpThpCalculator::
+getFloIPR(const WellState& well_state,
+          const Well& well, 
+          const SummaryState& summary_state) const 
+{
+    std::pair<double,double>retval(0.0, 0.0);
+    const auto& controls = well.productionControls(summary_state);
+    const auto& table = well_.vfpProperties()->getProd()->getTable(controls.vfp_table_number);
+    const auto& pu = well_.phaseUsage();
+    const auto& ipr_a= well_state.well(well_.indexOfWell()).implicit_ipr_a;
```

Comment:
Nitpicks: space around `=`, '?' and ':'.

Also, all of the `const auto&` here are just `const double`?